distributed/tests/test_cancelled_state.py

```diff
@@ -0,0 +1,219 @@
+import asyncio
+from unittest import mock
+
+import pytest
+
+from distributed import Nanny, wait
+from distributed.core import CommClosedError
+from distributed.utils_test import _LockedCommPool, gen_cluster, inc, slowinc
+
+pytestmark = pytest.mark.ci1
+
+
+@gen_cluster(client=True, nthreads=[("", 1)], Worker=Nanny)
```

Comment:
I think the Nanny was still a remnant of a previous test iteration and during cleanup I was hit by the same hang as you are. I think I mistakenly confused this with a loop problem due to scheduler, worker and client sharing the same loop. That's not true.

After some debugging, this test hangs because I am defining `wait_for_exec` using the future `fut` as a nonlocal variable. Upon serialization this creates another instance of the future, increasing the refcount on the client. For some reason (I guess our function serialization cache but I haven't confirmed) the ref for that additional instance is never decreased if I use a non-Nanny Worker. I don't have a good explanation why this doesn't happen with Nanny but simplifying the entire test case is resolving the issue and allows for better readability